README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# DTA Meta-analysis Sensitivity Analysis

<!-- badges: start -->

<!-- badges: end -->

[`dtametasa`](https://meta2020.github.io/dtametasa/) is a
likelihood-based sensitivity analysis for publication bias on SROC/SAUC
in meta-analysis of diagnostic studies

## Installation

You can install the released version from
[GitHub](https://github.com/meta2020/dtametasa) with:

``` r

# install.packages("devtools")
devtools::install_github("meta2020/dtametasa")
library(dtametasa)
```

## Data Format

Two formats of data are applicable:

  - Number of TP/FN/FP/TN

  - After logit transformation: y1/y2/v1/v2

<span style="color:red">**Attention: the column names of data must be
either of the above.**</span>

<!-- ## Example -->

<!-- This is an example which shows you how to solve a common problem.  -->

<!-- To take the data `IVD` as example, print the first several lines of data.  -->

<!-- ```{r} -->

<!-- ## Load package -->

<!-- library(dtametasa) -->

<!-- ## Load data -->

<!-- data(IVD) -->

<!-- kable(head(IVD)) -->

<!-- data(IVD2) -->

<!-- kable(head(IVD2)) -->

<!-- ``` -->

<!-- ### Main function 1: dtametasa.fc -->

<!-- This function need to pre-specify the c contrast in the selection function.  -->

<!-- ##### 1. Given a certain selection probability $p$, say, $p = 0.7$, we can get the estimation as follows. -->

<!-- ```{r} -->

<!-- ## Use default parameters setting -->

<!-- ## Print parameters' estimates -->

<!-- (sa1 <- dtametasa.fc(IVD, p = 0.7, b.interval = c(0, 2))) -->

<!-- ## If we change b.interval -->

<!-- (sa1 <- dtametasa.fc(IVD, p = 0.7, b.interval = c(0, 5))) -->

<!-- ## Use str() to get full results list -->

<!-- # str(sa1) -->

<!-- ``` -->

<!-- ##### 2. Given a series of selection probabilities, say, $p = 1, 0.9, 0,8, ...,0.1$.  -->

<!-- Attention: **$p$ must greater than 0 and cannot equal to 0. ($p>0$).** -->

<!-- ```{r} -->

<!-- ## Set p vectors -->

<!-- p.seq <- seq(1, 0.1, -0.1) -->

<!-- ## Get estimations for each p in p.seq vector -->

<!-- est1 <- sapply(p.seq, function(p) dtametasa.fc(IVD, p, b.interval = c(0, 2))$par.all) -->

<!-- ## Print estimation -->

<!-- colnames(est1)<- paste0("p = ", p.seq) -->

<!-- kable(est1) -->

<!-- ``` -->

<!-- ### Main function 2: dtametasa.rc -->

<!-- This function do not need to pre-specify the c contrast in the selection function.  -->

<!-- ##### 1. Given a certain selection probability $p$, say, $p = 0.7$, we can get the estimation as follows. -->

<!-- ```{r} -->

<!-- ## Use default parameters setting -->

<!-- ## Print parameters' estimates -->

<!-- (sa2 <- dtametasa.rc(IVD, p = 0.7)) -->

<!-- ## To get full results list -->

<!-- # str(sa2) -->

<!-- ``` -->

<!-- ##### 2. Given a series of selection probabilities, say, $p = 1, 0.9, 0,8, ...,0.1$.  -->

<!-- Attention: **$p$ must greater than 0 and cannot equal to 0. ($p>0$).** -->

<!-- ```{r} -->

<!-- ## Set p vectors -->

<!-- p.seq <- seq(1, 0.1, -0.1) -->

<!-- ## Get estimations for each p in p.seq vector -->

<!-- est2 <- sapply(p.seq, function(p) dtametasa.rc(IVD, p, b.interval = c(0, 2))$par.all) -->

<!-- ## Print estimation -->

<!-- colnames(est2)<- paste0("p = ", p.seq) -->

<!-- kable(est2) -->

<!-- ``` -->

<!-- ### Plot sroc -->

<!-- ##### 1. Single sroc -->

<!-- ```{r, fig.height=5, fig.width=10} -->

<!-- par(mfrow = c(1,2)) -->

<!-- ## This is the standard method: Reistma model -->

<!-- ## Without taking publication bias (PB) into consideration -->

<!-- library(mada) -->

<!-- fit <- reitsma(IVD, correction.control = "all", method = "ml") -->

<!-- plot(sroc(fit, type = "naive"), type = "l", ylim = c(0,1), xlim = c(0,1), col = "red") -->

<!-- ## Extact the estimation from Reistma model -->

<!-- par0 <- c(c(1,-1)*fit$coefficients, sqrt(fit$Psi[c(1,4)]), -fit$Psi[2]/prod(sqrt(fit$Psi[c(1,4)]))) -->

<!-- ## Add sroc -->

<!-- sroc.vec(sa1, add = TRUE, sroc.col = "black", sroc.lty = 1, spoint.pch = 1, spoint.col = "black") -->

<!-- with(IVD, points(FP/(FP+TN), TP/(TP+FN), pch = 4, cex = 0.5)) -->

<!-- legend("bottomright", c("Reistma", "dtametasa.fc", "IVD"),  -->

<!--        col = c("red", "black", "black"), lty = c(1,2, 0), pch = c(19,1, 4)) -->

<!-- title("When selection prob = 0.7, c1 = c2") -->

<!-- sroc.vec(par0, sroc.col = "red", spoint.col ="red") -->

<!-- ## Add sroc -->

<!-- sroc.vec(sa2, add = TRUE, sroc.col = "darkgray", sroc.lty = 1, spoint.col = "darkgray") -->

<!-- with(IVD, points(FP/(FP+TN), TP/(TP+FN), pch = 4, cex = 0.5)) -->

<!-- legend("bottomright", c("Reistma", "dtametasa.rc", "IVD"),  -->

<!--        col = c("red", "darkgray", "black"), lty = c(1, 2, 0), pch = c(19,19, 4)) -->

<!-- title("When selection prob = 0.7, estimate c1 c2") -->

<!-- par(mfrow = c(1,1)) -->

<!-- ``` -->

<!-- ##### 2. Multiple sroc -->

<!-- ```{r, fig.height=5, fig.width=10} -->

<!-- ## p vector and model -->

<!-- p.seq <- seq(1, 0.1, -0.1) -->

<!-- est1 <- sapply(p.seq, function(p) dtametasa.fc(IVD, p)$par) -->

<!-- est2 <- sapply(p.seq, function(p) dtametasa.rc(IVD, p)$par) -->

<!-- ## Plot multiple sroc -->

<!-- par(mfrow = c(1,2)) -->

<!-- sroc.mat(est1[1:5, ]) -->

<!-- sroc.vec(par0, add = TRUE, sroc.col = "red") -->

<!-- title("dtametasa.fc") -->

<!-- sroc.mat(est1[1:5, ]) -->

<!-- sroc.vec(par0, add = TRUE, sroc.col = "red") -->

<!-- title("dtametasa.rc") -->

<!-- par(mfrow = c(1,1)) -->

<!-- ``` -->

<!-- ### Calculate sAUC and confident interval -->

<!-- #### 1. Single sroc with CI -->

<!-- ```{r, fig.width=12, fig.height=6} -->

<!-- ## Use parameter vector -->

<!-- sa1 <- dtametasa.fc(IVD, p = 0.5, b.interval = c(0, 2)) -->

<!-- sa2 <- dtametasa.rc(IVD, p = 0.5, b.interval = c(0, 2)) -->

<!-- par(mfrow = c(1,2)) -->

<!-- sroc.vec(sa1) -->

<!-- title("dtametasa.fc") -->

<!-- sroc.vec(sa2) -->

<!-- title("dtametasa.rc") -->

<!-- par(mfrow = c(1,1)) -->

<!-- ``` -->

<!-- #### 2. Multiple sAUC -->

<!-- ```{r} -->

<!-- p.seq <- seq(1, 0.1, -0.1) -->

<!-- sauc1 <- sapply(p.seq, function(p) dtametasa.fc(IVD, p)$sauc.ci) -->

<!-- sauc2 <- sapply(p.seq, function(p) dtametasa.rc(IVD, p)$sauc.ci) -->

<!-- colnames(sauc1)<- paste0("p = ", p.seq) -->

<!-- kable(sauc1) -->

<!-- colnames(sauc2)<- paste0("p = ", p.seq) -->

<!-- kable(sauc2) -->

<!-- ``` -->

<!-- #### 4. Plot sAUC -->

<!-- This is an example of how to plot sAUC and CI. In analysis, please set `B = 1000` and then reproduce the plots. -->

<!-- ```{r, fig.height=5, fig.width=10} -->

<!-- par(mfrow = c(1,2)) -->

<!-- ## Use matplot to plot the sAUC and CI -->

<!-- matplot(t(sauc1), type = "b", lty = c(1,2,2),  -->

<!--         pch = 19, col = c("black", "grey", "grey"), -->

<!--         xlab = "p", ylab = "sAUC", -->

<!--         ylim = c(0,1), -->

<!--         xaxt = "n") -->

<!-- axis(1, at = 1:10, labels = p.seq) -->

<!-- title("dtametasa.fc") -->

<!-- matplot(t(sauc2), type = "b", lty = c(1,2,2),  -->

<!--         pch = 19, col = c("black", "grey", "grey"), -->

<!--         xlab = "p", ylab = "sAUC", -->

<!--         ylim = c(0,1), -->

<!--         xaxt = "n") -->

<!-- axis(1, at = 1:10, labels = p.seq) -->

<!-- title("dtametasa.rc") -->

<!-- par(mfrow = c(1,1)) -->

<!-- ``` -->


NEWS.md
# dtametasa 1.0.0

## 2021-01-18

- Add SROC and HSROC

- Update example data

# dtametasa 0.1.0

## 2021-09-10

- Update Description and Readme

- Update `sROC` and `sROC.matrix` functions into `sroc.vec` and `sroc.mat`

## 2021-05-26

- Update CI of SAUC

## 2021-03-26

- First release.


R/sAUC.ci.R
#' @title Parametric bootstrapped confidence interval of summary AUC
#'
#' @description Calculate the parametric bootstrap confidence interval (CI) for sAUC
#'
#' @param object object from function \code{dtametasa.fc} or \code{dtametasa.rc}.
#'
#' @param B The times for parametric bootstrapping.
#' Default is 1000.
#' It may be time consuming.
#'
#' @param ncores Set the an integer number of cores that will be used in parallel computing.
#' Default is 0, which detect the number of cores in the local device.
#'
#' @param type Computing types in the parallel computing.
#' See \code{type} augment in function \code{\link[parallel]{makeCluster}}.
#'
#' @param ci.level The significant value for confidence interval.
#' Default is 0.95, hence, a 2-tailed 95% CIs will be calculated by
#' profile likelihood.
#'
#' @param set.seed Give a integer to set the seed.
#' Default is NULL, which does not set any seed to fix the bootstrap results.
#'
#' @param hide.progress Whether to hide the progress bar in the calculation.
#' Default is not to hide.
#'
#' @return CI of sAUC
#'
#' @importFrom foreach foreach %dopar%
#' @importFrom doSNOW registerDoSNOW
#' @importFrom parallel makeCluster detectCores stopCluster
#' @importFrom stats quantile rnorm var
#' @importFrom utils setTxtProgressBar txtProgressBar install.packages
#' @importFrom doRNG %dorng%
#' @importFrom graphics matplot


#'
#' @seealso
#' \code{\link[parallel]{makeCluster}},
#'
#' @examples
#'
#' ## Here, we set B = 5 to save time in the calculation.
#' ## But, the results are not reliable.
#'
#' sa1 <- dtametasa.fc(IVD, p = 0.7)
#' (ci <- sAUC.ci(sa1, B = 5, set.seed = 1))
#'
#' @export

sAUC.ci <- function(object,
                    B = 1000,
                    ncores = 0, type = "SOCK",
                    ci.level = 0.95,
                    set.seed = NULL,
                    hide.progress = FALSE)
{
  if(!requireNamespace("foreach"))  install.packages("foreach")   else requireNamespace("foreach")
  if(!requireNamespace("parallel")) install.packages("parallel")  else requireNamespace("parallel")
  if(!requireNamespace("doSNOW"))   install.packages("doSNOW")    else requireNamespace("doSNOW")
  if(!requireNamespace("doRNG"))    install.packages("doRNG")     else requireNamespace("doRNG")

  if(!inherits(object, "dtametasa")) stop("ONLY VALID FOR RESULTS OF dtametasa.fc OR dtametasa.rc")

  data <- object$data

  S  <- nrow(data)
  y1 <- data$y1
  y2 <- data$y2
  v1 <- data$v1
  v2 <- data$v2

  ## >>
  u <- object$par[c(1,2)]

  SO <- lapply(1:S, function(s){

    matrix(c(v1[s], 0, 0, v2[s]),2,2) + matrix(object$par[c(3,5,5,4)], 2,2)

  })

  ## <<

  if(ncores == 0) ncores <- detectCores() else ncores <- ceiling(ncores)

  cl <- makeCluster(ncores, type = type)
  registerDoSNOW(cl)

  opts <- NULL

  if (!hide.progress){

    pb <- txtProgressBar(max = B, style = 3, width = 40)
    progress <- function(n) setTxtProgressBar(pb, n)
    opts <- list(progress = progress)
  }


  if(object$func.name == "dtametasa.fc"){

    set.seed(set.seed)
    par <- foreach(r=1:B, .combine=rbind, .packages = "dtametasa", .options.snow = opts)  %dorng%  {

      # y1.t <- sapply(1:S, function(i) rnorm(1,y1[i],sqrt(v1[i])))
      # y2.t <- sapply(1:S, function(i) rnorm(1,y2[i],sqrt(v2[i])))
      # data.t <- data.frame(y1 = y1.t, y2 = y2.t, v1 = v1, v2 = v2)

      Y <- t(sapply(1:S, function(i) mvtnorm::rmvnorm(1, u, SO[[i]])))

      data.t <- data.frame(y1 = Y[,1], y2 = Y[,2], v1 = v1, v2 = v2)

      args <- c(list(data = data.t), object$pars.info)
      sa1 <- try(do.call("dtametasa.fc", args), silent = TRUE)
      if (!inherits(sa1,"try-error")) sa1$par[c(1,2,4,5, 10)] else rep(NA, 5)

    }
  }

  if(object$func.name == "dtametasa.rc"){

    set.seed(set.seed)

    par <- foreach(r=1:B, .combine=rbind, .packages = "dtametasa", .options.snow = opts)  %dorng%  {

      # y1.t <- sapply(1:S, function(i) rnorm(1,y1[i], sqrt(v1[i])))
      # y2.t <- sapply(1:S, function(i) rnorm(1,y2[i], sqrt(v2[i])))
      # data.t <- data.frame(y1 = y1.t, y2 = y2.t, v1 = v1, v2 = v2)

      Y <- t(sapply(1:S, function(i) mvtnorm::rmvnorm(1, u, SO[[i]])))

      data.t <- data.frame(y1 = Y[,1], y2 = Y[,2], v1 = v1, v2 = v2)

      args <- c(list(data = data.t), object$pars.info)
      sa2 <- try(do.call("dtametasa.rc", args), silent = FALSE)
      if (!inherits(sa2,"try-error")) sa2$par[c(1,2,4,5, 10)] else rep(NA, 5)

  }
  }

  if (!hide.progress) close(pb)
  stopCluster(cl)

  PAR <- as.matrix(par)
  sauc.t   <- PAR[, 5]

  # n <- length(sauc.t)
  # se <- (n-1)/n * sd(sauc.t, na.rm = TRUE)

  n <- sum(!is.na(sauc.t))
  se <- sqrt((n-1)/n * var(sauc.t, na.rm = TRUE))

  sl.sauc.t <- (sauc.t - mean(sauc.t, na.rm = TRUE))/se

  # s.sauc.r <- order(sl.sauc.t)
  #PAR.r <- PAR[, s.sauc.r]

  # s.sauc.t <- sort(sl.sauc.t)

  q1 <- quantile(sl.sauc.t, probs = (1-ci.level)/2, na.rm = TRUE)
  q2 <- quantile(sl.sauc.t, probs = 1-(1-ci.level)/2, na.rm = TRUE)

  sauc <- object$par[10]

  list <- list(sauc = sauc,
       ci.l = max(sauc+q1*se, 0),
       ci.u = min(sauc+q2*se, 1),
       bootstrap.par  = PAR,
       cluster = cl)


  class(list) <- "sAUC.ci"

  list

}

#' @title Print sAUC.ci results
#'
#' @description Print results from function \code{\link{sAUC.ci}}
#'
#' @param x object from function \code{\link{sAUC.ci}}.
#'
#' @param digits digits of the results.
#'
#' @param ... other parameters in function \code{\link{print}}.
#'
#' @seealso
#' \code{\link{sAUC.ci}};
#' \code{\link[base]{print}}
#'
#' @rdname print.sAUC.ci
#'
#' @export
#'
print.sAUC.ci <- function(x, digits = 3, ...){

  if(!inherits(x, "sAUC.ci")) stop("ONLY VALID FOR RESULTS OF sAUC.ci")

  sauc <- c(x$sauc, x$ci.l, x$ci.u)
  names(sauc) <- c("sauc", "ci.l", "ci.u")

  print(sauc, digits = digits, ...)

}

#' @title Plot sROC.ci results
#'
#' @description Plot the sROC from function \code{\link{sAUC.ci}}
#'
#' @param x object from function \code{\link{sAUC.ci}}.
#'
#' @param sroc.ci.col The color of the CIs of sROC.
#' Default is grey.
#'
#' @param sroc.ci.lty The line type of CIs.
#' Default is solid line.
#'
#' @param sroc.ci.lwd The line width of CIs.
#' Defalt is 1.
#'
#' @param add.sROC.ci Whether to add the plot of sROC with the CIs lines
#' onto a new plot.
#' Default is not to add.
#'
#' @param ci.level The significant value for confidence interval.
#' Default is 0.95, hence, a 2-tailed 95% CIs will be calculated by
#' profile likelihood.
#'
#' @param ... Other augments in function \code{\link{sROC}}
#'
#' @seealso
#' \code{\link{sAUC.ci}};
#' \code{\link[graphics]{matplot}},
#'
#' @rdname plot.sAUC.ci
#'
#' @examples
#'
#' sa1 <- dtametasa.fc(IVD, p = 0.7)
#'
#' sROC(sa1)
#' ci <- sAUC.ci(sa1, B = 5, set.seed = 1)
#' plot(ci)
#'
#' (p <- plot(ci))
#'
#' @export
#'

plot.sAUC.ci <- function(x,
                         sroc.ci.col = "grey",
                         sroc.ci.lty = 1,
                         sroc.ci.lwd = 1,
                         add.sROC.ci=TRUE,
                         ci.level = 0.95,
                         ...){

  if(!inherits(x, "sAUC.ci")) stop("ONLY VALID FOR RESULTS OF sAUC.ci")

  PAR <- x$bootstrap.par
  B <- dim(PAR)[1]


  fpr.t <- seq(0,1,0.001)
  se.t  <- sapply(1:B, function(i){

    u1  <- PAR[i,1]
    u2  <- PAR[i,2]
    t22 <- PAR[i,3]
    t12 <- PAR[i,4]

    plogis(u1 - (t12/t22) * (qlogis(fpr.t) + u2))

  })


  #sROC(object, add.spoint = add.spoint, add = add.plot.ROC.ci, ...)

  ci <- cbind(

    apply(se.t, 1, function(x) quantile(x, (1-ci.level)/2, na.rm = TRUE)),
    apply(se.t, 1, function(x) quantile(x, probs = 1-(1-ci.level)/2, na.rm = TRUE))

  )

  matplot(x = fpr.t, y = ci, type = "l",
          col = sroc.ci.col, lty = sroc.ci.lty, lwd = sroc.ci.lty,
          add = TRUE)

  list <- c(x, sroc.ci <- list(x = fpr.t, y = ci))


}
